stream/stream_dsm.c

```diff
@@ -0,0 +1,472 @@
+/*
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+
+#include <bdsm/bdsm.h>
+#include <unistd.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <netdb.h>
+
+#include "common/msg.h"
+#include "stream.h"
+#include "options/m_option.h"
+
+#include "config.h"
+
+static pthread_mutex_t dsm_lock = PTHREAD_MUTEX_INITIALIZER;
+
+struct priv {
+    char *domain;
+    char *user;
+    char *password;
+    char *server;
+    char *share;
+    char *filepath;
+    smb_session *session;
+    smb_tid shareID;
+    smb_fd fd;
+};
+
+static char** str_split(char* str, const char delim_)
```

Comment:
Any reason other than not utilizing the bstr helpers like `bstr_split`? Requires bstr'ification, but these functions are otherwise utilized around the code base. See [misc/bstr.h](https://github.com/mpv-player/mpv/blob/master/misc/bstr.h) for details.

Reply:
Thanks, I'll try implement the string part by bstr.h.
libdsm didn't provide a protocol url parser like libsmbclient does.
The smb:// url is really something complex for C lang.